math/vecops/test/vecops_rvec.cxx

```diff
@@ -1191,3 +1088,95 @@ TEST(VecOps, Construct)
    EXPECT_TRUE(fourVects[2] == ref2);
 }
 
+bool IsSmall(const RVec<int> &v)
+{
+   // the first array element is right after the 3 data members of SmallVectorBase
+   return reinterpret_cast<std::uintptr_t>(v.begin()) - reinterpret_cast<std::uintptr_t>(&v) ==
+          sizeof(void *) + 2 * sizeof(int);
+}
+
+// this is a regression test for https://github.com/root-project/root/issues/6796
+TEST(VecOps, MemoryAdoptionAndClear)
+{
+   ROOT::RVec<int> v{1, 2, 3};
+   EXPECT_TRUE(IsSmall(v));
+   ROOT::RVec<int> v2(v.data(), v.size());
+   v2[0] = 0;
+   v2.clear();
+   EXPECT_TRUE(All(v == RVec<int>{0, 2, 3}));
+   v2.push_back(42);
+   EXPECT_FALSE(IsSmall(v2)); // currently RVec does not go back to a small state after `clear()`
+   EXPECT_TRUE(All(v2 == RVec<int>{42}));
+}
+
+// interaction between small buffer optimization and memory adoption
+TEST(VecOps, MemoryAdoptionAndSBO)
+{
+   int *values = new int[3]{1, 2, 3};
+   ROOT::RVec<int> v(values, 3);
+   auto check = [](const RVec<int> &mv) {
+      EXPECT_EQ(mv.size(), 3u);
+      EXPECT_EQ(mv[0], 1);
+      EXPECT_EQ(mv[1], 2);
+      EXPECT_EQ(mv[2], 3);
+   };
+   check(v);
+   EXPECT_FALSE(IsSmall(v));
+   ROOT::RVec<int> v2 = std::move(v);
+   EXPECT_TRUE(v.empty());
+   check(v2);
+   // this changes the RVec from memory adoption mode to "long" mode, even if the size is small
+   // currently we don't allow going from memory adoption to small buffer mode directly, it could be future optimization
+   v2.push_back(4);
+   EXPECT_FALSE(IsSmall(v2));
+   v2.clear();
+   v2.push_back(1);
+   v2.push_back(2);
+   v2.push_back(3);
+   check(v2);
+   delete[] values;
+   check(v2);
+}
+
+struct ThrowingCtor {
+   ThrowingCtor() { throw std::runtime_error("This exception should have been caught."); }
+};
+
+struct ThrowingMove {
+   ThrowingMove() {}
+   ThrowingMove(const ThrowingMove &) {}
+   ThrowingMove &operator=(const ThrowingMove &) { return *this; }
+   ThrowingMove(ThrowingMove &&) { throw std::runtime_error("This exception should have been caught."); }
+   ThrowingMove &operator=(const ThrowingMove &&) { return *this; }
+};
+
+// RVec does not guarantee exception safety, but we still want to test
+// that we don't segfault or otherwise crash if element construction or move throws.
+TEST(VecOps, NoExceptionSafety)
```

Comment:
Could you add an additional test for a throwing copy constructor or a copy assignment operator? I believe this is more common than a throwing move ctor.

Reply:
done, let me know if you'd add something else